sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java

```diff
@@ -0,0 +1,186 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.redis;
+
+import static org.junit.Assert.assertEquals;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.RunnableOnService;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Test on the Redis IO.
+ */
+public class RedisIOTest {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RedisIOTest.class);
+
+  @Rule public TestPipeline pipeline = TestPipeline.create();
+
+  @Test
+  public void testEstimatedSizeBytes() throws Exception {
+    final FakeRedisService service = new FakeRedisService();
+    service.clear();
+    for (int i = 0; i < 10; i++) {
+      service.put("Foo " + i, "Bar " + i);
+    }
+
+    PipelineOptions pipelineOptions = PipelineOptionsFactory.create();
+    RedisIO.RedisSource source = new RedisIO.RedisSource("Foo.*",
+        new SerializableFunction<PipelineOptions, RedisService>() {
+          @Override
+          public RedisService apply(PipelineOptions input) {
+            return service;
+          }
+        });
+    long estimatedSizeBytes = source.getEstimatedSizeBytes(pipelineOptions);
+    LOG.info("Estimated size: {}", estimatedSizeBytes);
+    assertEquals(100, estimatedSizeBytes);
+  }
+
+  @Test
+  @Category(RunnableOnService.class)
+  public void testGetRead() throws Exception {
+    FakeRedisService service = new FakeRedisService();
+    for (int i = 0; i < 1000; i++) {
+      service.put("Key " + i, "Value " + i);
+    }
+
+    PCollection<KV<String, String>> output =
+        pipeline.apply(RedisIO.read().withRedisService(service));
+
+    PAssert.thatSingleton(output.apply("Count", Count.<KV<String, String>>globally()))
+        .isEqualTo(1000L);
+    ArrayList<KV<String, String>> expected = new ArrayList<>();
```

Comment:
You can put this at the top of the test, and at line 78 instead iterate over `expected` to avoid effectively duplicating the code describing the expected data.